### PR TITLE
add: option to force lowercase cname

### PR DESCRIPTION
**Pull request by ziirish**
Hello,

As already discussed, when you are evolving in an heterogeneous environment (with MacOS, Windows and Linux clients), and you are using the default `cname` (returned by get_fqdn) you may end-up with mixed case `cname`s (Windows likes uppercase...).

This patch allows to force lowercase `cname` with the new option `cname_lowercase` which defaults to `0`.
The option can be set globally on the server or client-side.

### CODE REVIEW

**Comment by ziirish** [comment]
Note: There is a second patch I'd like to submit but it depends on this one

For the same needs, I'd like to strip FQDN because MacOS returns `hostname.local`, Windows returns `HOSTNAME` and Linux returns `hostname.f.q.d.n`

File: src/server/auth.c [hunk 1]
```diff
@@ -153,7 +154,12 @@ int authorise_server(struct asfd *asfd,
 		logp("unable to get client name\n");
 		goto end;
 	}
-	if(set_string(cconfs[OPT_CNAME], rbuf->buf))
+	cname=strdup(rbuf->buf);
```
Comment (grke) [inline src/server/auth.c]: This should be strdup_w(rbuf->buf, __func__).
Something should check that the result is not NULL:
if(!(cname=strdup_w(rbuf->buf, __func__)))
  goto end;

File: src/server/auth.c [hunk 2]
```diff
@@ -184,5 +190,6 @@ int authorise_server(struct asfd *asfd,
 end:
 	iobuf_free_content(rbuf);
 	free_w(&password);
+	if(cname) free(cname);
```
Comment (grke) [inline src/server/auth.c]: This should be free_w(&cname);

File: utest/server/test_auth.c [hunk 3]
```diff
@@ -232,6 +263,9 @@ START_TEST(test_authorise_server)
 	do_test(-1, setup_no_password_configured);
 	do_test(-1, setup_passwd_failed);
 	do_test(-1, setup_no_keep_configured);
+	do_test(-1, setup_lower_failed);
+	FORCE_LOWERCASE=1;
```
Comment (grke) [inline utest/server/test_auth.c]: Can this be passed into do_test as a parameter instead?

File: utest/test_conf.c [hunk 4]
```diff
@@ -91,6 +91,7 @@ static void check_default(struct conf **c, enum conf_opt o)
 		case OPT_S_SCRIPT_POST_NOTIFY:
 		case OPT_S_SCRIPT_NOTIFY:
 		case OPT_HARDLINKED_ARCHIVE:
+			// wrong indent?
```
Comment (grke) [inline utest/test_conf.c]: It has spaces instead of tabs, by accident.

File: src/conffile.c [hunk 5]
```diff
@@ -621,13 +622,17 @@ static int get_fqdn(struct conf **c)
 		goto end;
 	}
 
-	if(set_string(c[OPT_CNAME], info->ai_canonname))
+	fqdn=strdup(info->ai_canonname);
```
Comment (grke) [inline src/conffile.c]: Needs to be strdup_w, and check that the result is not NULL:
if(!(fqdn=strdup_w(info->ai_canonname, __func__)))
  goto end;

File: src/conffile.c [hunk 6]
```diff
@@ -621,13 +622,17 @@ static int get_fqdn(struct conf **c)
...
+	if(get_int(c[OPT_CNAME_LOWERCASE])) strlwr(fqdn);
+
+	if(set_string(c[OPT_CNAME], fqdn))
 		goto end;
 	logp("cname from hostname: %s\n", get_string(c[OPT_CNAME]));
 
 	ret=0;
 end:
 	if(info) freeaddrinfo(info);
+	if(fqdn) free(fqdn);
```
Comment (grke) [inline src/conffile.c]: free_w(&fqdn);

**Comment by grke** [comment]
Thanks for this.
I put a bunch of comments on the patches.
I also fixed travis on master so that OSX is passing again. If you rebase on master, you should get green ticks everywhere.

**Comment by ziirish** [comment]
Hi,

Thanks for reviewing this.

I'm fixing all this and I'll update the pull request asap.

**Comment by grke** [comment]
Hello,
Sorry for taking ages to get back to this.
Something is failing in travis related to this patch on OSX:

99%: Checks: 314, Failures: 1, Errors: 0
../../utest/builders/build_asfd_mock.c:55:F:Core:test_authorise_server:0: Assertion 'writes->cursor<writes->size' failed

**Comment by ziirish** [comment]
No problem, we are all busy ;-)

I noticed the OSX error, unfortunately I have no such environment @home to try to debug.

I had this error previously on Linux too due to a missing statement in the `setup_lower_ok` test. The missing line was: `asfd_assert_write(asfd, &w, 0, CMD_GEN, "ok");`

**Comment by grke** [comment]
OK, I will try to set up the VM that somebody sent me a while back, and figure out what is going on.
But, it might be a few days.

**Comment by ziirish** [comment]
Well don't worry, I'll run some tests against this branch, but this is probably going to be noisy.

**Comment by grke** [comment]
OK, I am prepared. :)

**Comment by grke** [comment]
I figured it out.

These stupid Macs have case-insensitive file systems, which means that your 'setup_lower_failed' test fails because it always finds the file 'clientconfdir/testclient', even though burp asked for 'clientconfdir/TESTCLIENT'.

I think the best thing to do is to ifndef HAVE_DARWIN_OS around that test.

If you were to use a Mac as a burp server, this could have some horrible implications.

**Comment by ziirish** [comment]
Wow, nice one you found here!

I'll patch the test.
Thank you for the debug and explanations.

**Comment by grke** [comment]
Nice job. I will merge it when I next have a bit of spare time. Should be pretty soon.

**Pull request merged**